inference/primitive.go

```diff
@@ -33,72 +37,212 @@ import (
+	}
+
+	producer, consumer := trigger.Prestrings(p.pass)
+	return primitiveFullTrigger{
+		Position:     p.toPosition(trigger.Consumer.Expr.Pos()),
+		ProducerRepr: producer,
+		ConsumerRepr: consumer,
+	}
 }
 
-// newPrimitiveSite encodes a passed Key as a primitiveSite, needing a boolean `isDeep` to complete
-// the translation.
-// As discussed above for `primitiveSite`, it is vital that this function is injective,
-// deterministic, and produces minimized output. Multi Package inference relies on all of these
-// properties.
-func newPrimitiveSite(key annotation.Key, isDeep bool) primitiveSite {
-	pkgStringRepr := ""
+// site returns the primitive version of the annotation site.
+func (p *primitivizer) site(key annotation.Key, isDeep bool) primitiveSite {
+	objPath, err := objectpath.For(key.Object())
+	if err != nil {
+		// An error will occur when trying to get object path for unexported objects, in which case
+		// we simply assign an empty object path.
+		objPath = ""
+	}
+
+	pkgRepr := ""
 	if pkg := key.Object().Pkg(); pkg != nil {
-		pkgStringRepr = pkg.Path()
+		pkgRepr = pkg.Path()
+	}
+
+	var position token.Position
+	// For upstream objects, we need to look up the local position cache for correct positions.
+	if key.Object().Pkg() != p.pass.Pkg {
+		// Correct upstream information may not always be in the cache: we may not even have it
+		// since we skipped analysis for standard and 3rd party libraries.
+		if p, ok := p.upstreamObjPositions[pkgRepr+"."+string(objPath)]; ok {
+			position = p
+		}
+	}
+
+	// Default case (local objects or objects from skipped upstream packages), we can simply use
+	// their Object.Pos() and retrieve the position information. However, we must trim the possible
+	// build-system sandbox prefix from the filenames for cross-package references.
+	if !position.IsValid() {
+		position = p.toPosition(key.Object().Pos())
 	}
+
 	return primitiveSite{
-		StringRepr:    key.String(),
-		IsDeep:        isDeep,
-		Pos:           key.Object().Pos(),
-		PkgStringRepr: pkgStringRepr,
-		Exported:      key.Object().Exported(),
+		PkgPath:    pkgRepr,
+		Repr:       key.String(),
+		IsDeep:     isDeep,
+		Exported:   key.Object().Exported(),
+		ObjectPath: objPath,
+		Position:   position,
 	}
 }
 
-func (s primitiveSite) String() string {
-	deepStr := ""
-	if s.IsDeep {
-		deepStr = "Deep "
+// toPosition returns the correct position information for the given pos, removing sandbox prefix
+// if any.
+func (p *primitivizer) toPosition(pos token.Pos) token.Position {
+	// Generated files contain "//line" directives that point back to the original source file
+	// for better error reporting, and PositionFor supports reading that information and adjust
+	// the position accordingly. However, those source files are never truly analyzed, meaning
+	// the downstream analysis will try to look for the generated files instead of the source
+	// files. Therefore, here we use the unadjusted position instead.
```

Comment:
I get an idea of what this comment is about, but not quite the consequences of passing `adjusted=true` here instead... what would that cause? (Just out of curiosity. I assume the choice is correct, but the explanation isn't super easy to follow for me now that I forgot so much of the context of our discussions about this PR :) )